okhttp/src/test/java/okhttp3/CacheCorruptionTest.kt

```diff
@@ -0,0 +1,109 @@
+import java.net.ResponseCache
+import java.text.DateFormat
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+import java.util.TimeZone
+import java.util.concurrent.TimeUnit
+import javax.net.ssl.HostnameVerifier
+import javax.net.ssl.SSLSession
+
+class CacheCorruptionTest {
+  @get:Rule
+  var server = MockWebServer()
+
+  @get:Rule
+  var fileSystem = InMemoryFileSystem()
+
+  @get:Rule
+  val clientTestRule = OkHttpClientTestRule()
+
+  @get:Rule
+  val platform = PlatformRule()
+
+  private val handshakeCertificates = localhost()
+  private lateinit var client: OkHttpClient
+  private lateinit var cache: Cache
+  private val NULL_HOSTNAME_VERIFIER =
+    HostnameVerifier { name: String?, session: SSLSession? -> true }
+  private val cookieManager = CookieManager()
+
+  @Before fun setUp() {
+    platform.assumeNotOpenJSSE()
+    platform.assumeNotBouncyCastle()
+    server.protocolNegotiationEnabled = false
+    cache = Cache(File("/cache/"), Int.MAX_VALUE.toLong(), fileSystem)
+    client = clientTestRule.newClientBuilder()
+      .cache(cache)
+      .cookieJar(JavaNetCookieJar(cookieManager))
+      .build()
+  }
+
+  @After fun tearDown() {
+    ResponseCache.setDefault(null)
+    if (this::cache.isInitialized) {
+      cache.delete()
+    }
+  }
+
+  @Test @Throws(IOException::class) fun corruptedCacheEntry() {
+    server.useHttps(handshakeCertificates.sslSocketFactory(), false)
+    server.enqueue(MockResponse()
+      .addHeader("Last-Modified: " + formatDate(-1, TimeUnit.HOURS))
+      .addHeader("Expires: " + formatDate(1, TimeUnit.HOURS))
+      .setBody("ABC"))
+    client = client.newBuilder()
+      .sslSocketFactory(
+        handshakeCertificates.sslSocketFactory(), handshakeCertificates.trustManager)
+      .hostnameVerifier(NULL_HOSTNAME_VERIFIER)
+      .build()
+    val request: Request = Request.Builder().url(server.url("/")).build()
+    val response1: Response = client.newCall(request).execute()
+    val bodySource = response1.body!!.source()
+    Assertions.assertThat(bodySource.readUtf8()).isEqualTo("ABC")
+
+    val metadataFile = fileSystem.files.keys.find { it.name.endsWith(".0") }
+    val metadataBuffer = fileSystem.files[metadataFile]
+
+    val contents = metadataBuffer!!.peek().readUtf8()
+
+    metadataBuffer.writeUtf8(contents.substring(0, contents.length / 4))
+
+//    println(metadataBuffer.peek().readUtf8())
+
+    val response2: Response = client.newCall(request).execute() // Cached!
+    Assertions.assertThat(response2.body!!.string()).isEqualTo("ABC")
+    Assertions.assertThat(cache.requestCount()).isEqualTo(2)
+    Assertions.assertThat(cache.networkCount()).isEqualTo(1)
+    Assertions.assertThat(cache.hitCount()).isEqualTo(1)
+  }
+
+  /**
+   * @param delta the offset from the current date to use. Negative values yield dates in the past;
+   * positive values yield dates in the future.
+   */
+  private fun formatDate(delta: Long, timeUnit: TimeUnit): String? {
```

Comment:
this is handy

Reply:
copy and pasted from the larger cache test.  I have problems dealing with massive source files, they break my brain so I added this test file to cover any fuzzing or deliberate break the cache scenarios.